File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.go
func letterCombinations(digits string) []string {

	f:=map[byte]string{
		'2':"abc",
		'3':"def",
		'4':"ghi",
		'5':"jkl",
		'6':"mno",
		'7':"pqrs",
		'8':"tuv",
		'9':"wxyz",
	}

	n:=len(digits)
	diary:=""
	res:=[]string{}

	if n==0{
		return res
	}

	var fun func(string,int,int,*string,[]string) []string

	fun=func(digits string,n int,idx int,diary *string,res []string) []string{

		// Last wla stage lelo
		if idx==n{
			res=append(res,*diary)
			return res
		}

		// Strings ko button ke nichey hai sab ek sath nilega
		choice:=f[digits[idx]]

		for j:=0;j<len(choice);j++{

			// Push kardo har j k index ko
			*diary+=string(choice[j])

			// Explore karlo next wley mei jakey
			res=fun(digits,n,idx+1,diary,res)

			// Reverse kardo
			*diary=(*diary)[:len(*diary)-1]
		}

		return res
	}

	// Recursive Call kardo intio main
	res=fun(digits,n,0,&diary,res)

	return res
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def fun(self,digits,n,idx,diary,res,f):

        # Last wla stage lelo
        if idx==n:
            res.append("".join(diary))      # res hi mein push back kardo diary wla
            return

        # Strings ko button ke nichey hai sab ek sath nilega
        choice=f[digits[idx]]

        for j in range(len(choice)):

            # Push kardo har j k index ko
            diary.append(choice[j])

            # Explore karlo next wley mei jakey
            self.fun(digits,n,idx+1,diary,res,f)

            # Reverse kardo
            diary.pop()

        return

    def letterCombinations(self,digits):
        f={}

        # HashMap dec kardo taki sab ek sath ajae(1-9 tak)
        f['2']="abc"
        f['3']="def"
        f['4']="ghi"
        f['5']="jkl"
        f['6']="mno"
        f['7']="pqrs"
        f['8']="tuv"
        f['9']="wxyz"

        n=len(digits)
        diary=[]               # shuru mein empty hi hoga
        res=[]                 # Result k liye

        if n==0:
            return res

        # Recursive Call kardo intio main
        self.fun(digits,n,0,diary,res,f)

        return res